File: app/src/androidTest/java/sample/architecture_components/persistence/WordDaoTest.java
package sample.architecture_components.persistence;

import android.arch.core.executor.testing.InstantTaskExecutorRule;

import android.arch.persistence.room.Room;
import android.content.Context;
import android.support.test.InstrumentationRegistry;
import android.support.test.runner.AndroidJUnit4;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.ArrayList;
import java.util.List;

import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;

/**
 * Instrumented test, which will execute on an Android device.
 *
 * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
 */
@RunWith(AndroidJUnit4.class)
public class WordDaoTest {
    @Rule
    public InstantTaskExecutorRule instantTaskExecutorRule = new InstantTaskExecutorRule();

    private WordDao wordDao;
    private WordRoomDatabase db;

    @Before
    public void createDb() {
        Context appContext = InstrumentationRegistry.getTargetContext();
        // Using an in-memory database because the information stored here disappears when the
        // process is killed.
        db = Room.inMemoryDatabaseBuilder(appContext, WordRoomDatabase.class)
                // Allowing main thread queries, just for testing.
                .allowMainThreadQueries()
                .build();
        wordDao = db.wordDao();
    }

    @After
    public void closeDb() {
        db.close();
    }

    @Test
    public void testInsert() throws InterruptedException {
        Word word = new Word("test");
        wordDao.insert(word);
        List<Word> words = TestUtil.getData(wordDao.getAllWords());
        assertThat(words.get(0).getWord(), is(word.getWord()));
    }

    @Test
    public void testGetAll() throws InterruptedException {
        final Word word1 = new Word("test1");
        wordDao.insert(word1);
        final Word word2 = new Word("test2");
        wordDao.insert(word2);
        List<Word> expectedWords = new ArrayList<Word>() {{
            add(word1);
            add(word2);
        }};

        List<Word> words = TestUtil.getData(wordDao.getAllWords());
        assertThat(words, is(equalTo(expectedWords)));
    }

    @Test
    public void testDeleteAll() throws InterruptedException {
        // Given
        final Word word1 = new Word("test1");
        wordDao.insert(word1);
        final Word word2 = new Word("test2");
        wordDao.insert(word2);
        // When
        wordDao.deleteAll();
        // Then
        List<Word> words = TestUtil.getData(wordDao.getAllWords());
        assertThat(words.isEmpty(), is(true));
    }

}

File: app/src/androidTest/java/sample/architecture_components/persistence/WordDaoTest.kt
package sample.architecture_components.persistence

import android.app.Application
import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import androidx.room.Room
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4

import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

import java.util.ArrayList

import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.MatcherAssert.assertThat
//import sample.architecture_components.AndroidRoomSample

@RunWith(AndroidJUnit4::class)
class WordDaoTest {
    @get:Rule
    val instantTaskExecutorRule = InstantTaskExecutorRule()

    private lateinit var wordDao: WordDao
    private lateinit var db: WordRoomDatabase

    @Before
    fun createDb() {
        val appContext = ApplicationProvider.getApplicationContext<Application>()
        // Using an in-memory database because the information stored here disappears when the
        // process is killed.
        db = Room.inMemoryDatabaseBuilder(appContext, WordRoomDatabase::class.java)
                // Allowing main thread queries, just for testing.
                .allowMainThreadQueries()
                .build()
        wordDao = db.wordDao()
    }

    @After
    fun closeDb() {
        db.close()
    }

    @Test
    @Throws(InterruptedException::class)
    fun testInsert() {
        val word = Word("test")
        wordDao.insert(word)
        val words = TestUtil.getData(wordDao.getAllWords())
        assertThat(words[0].word, `is`(word.word))
    }

    @Test
    @Throws(InterruptedException::class)
    fun testGetAll() {
        val word1 = Word("test1")
        wordDao.insert(word1)
        val word2 = Word("test2")
        wordDao.insert(word2)
        val expectedWords = object : ArrayList<Word>() {
            init {
                add(word1)
                add(word2)
            }
        }

        val words = TestUtil.getData(wordDao.getAllWords())
        assertThat(words, `is`(equalTo<List<Word>>(expectedWords)))
    }

    @Test
    @Throws(InterruptedException::class)
    fun testDeleteAll() {
        // Given
        val word1 = Word("test1")
        wordDao.insert(word1)
        val word2 = Word("test2")
        wordDao.insert(word2)
        // When
        wordDao.deleteAll()
        // Then
        val words = TestUtil.getData(wordDao.getAllWords())
        assertThat(words.isEmpty(), `is`(true))
    }

}